doc/docs.md

```diff
@@ -3662,6 +3663,250 @@ fn test() []int {
+the compiler would complain about the assignment in `f()` because `s` *"might
+refer to an object stored on stack"*. The assumption made in `g()` that the call
+`r.g(&s)` would only borrow the reference to `s` is wrong. 
+
+A solution to this dilemma is the `[heap]` attribute at the declaration of
+`struct MyStruct`. It instructs the compiler to *always* allocate `MyStruct`-objects
+on the heap. This way the references to `s` remains valid even after `g()` returns.
+The compiler takes into consideration that `MyStruct` objects are always heap
+allocated when checking `f()` and allows assigning the reference to `s` to the
+`r.r` field.
+
+There is a pattern often seen in other programming languages:
+
+```v failcompile
+fn (mut a MyStruct) f() &MyStruct {
+	// do something with a
+	return &a // would return address of borrowed object
+}
+```
+
+Here `f()` is passed a reference `a` that is passed back to the caller and returned
+at the same time. The intention behind such a declaration is method chaining like
+`y = x.f().g()`. However, the problem with this approach is that a second reference
+to `a` is created &ndash; so it is not only borrowed and `MyStruct` has to be
+declared as `[heap]`.
+
+In V the better approach is:
+
+```v
+struct MyStruct {
+mut:
+	n int
+}
+
+fn (mut a MyStruct) f() {
+	// do something with `a`
+}
+
+fn (mut a MyStruct) g() {
+	// do something else with `a`
+}
+
+fn main() {
+	x := MyStruct{} // stack allocated
+	mut y := x
+	y.f()
+	y.g()
+	// instead of `mut y := x.f().g()
+}
+```
+
+This way the `[heap]` attribute can be avoided &ndash; resulting in better performance.
+
+However, stack space is very limited as mentioned above. For this reason the `[heap]`
+attribute might be suitable for very large structures even if not required by use cases
+like those mentioned above.
+
+There is an alternative way to manually control allocation on a case to case basis:
+
+```v
+struct RefStruct {
+mut:
+	r &MyStruct
+}
+
+struct MyStruct {
+	n int
+}
+
+fn (mut r RefStruct) f(s &MyStruct) {
+	r.r = unsafe { s } // override compiler check
+}
+
+fn (mut r RefStruct) g() {
+	s := &MyStruct{ // `s` explicitly referes to a heap object
+		n: 7
+	}
+	r.f(s)
+}
+
+fn use_stack() {
+	x := 7.5
+	y := 3.25
+	z := x + y
+	println('$x $y $z')
+}
+
+fn main() {
+	m := MyStruct{}
+	mut r := RefStruct{
+		r: &m
+	}
+	r.g()
+	use_stack() // to erase invalid stack contents
+	println('r: $r')
+}
+```
+
+Here the compiler check is suppressed by the `unsafe` block. To make `s` be heap
+allocated even without `[heap]` attribute the `struct` literal is prefixed with
+an ampersand: `&MyStruct{...}`.
+
+This last step would not be required by the compiler but without it the reference
+inside `r` becomes invalid (the memory area pointed to will be overwritten by
+`use_stack()`) and the program might crash (or at least produce an unpredictable
+final output). That's why this approach is *unsafe*!
```

Comment:
I am quite confused by this, so is this a negative example of how not to do things?
Sounds like the proposed code snippet above should be avoided.

Reply:
Sorry for the multiple comments... Should have started a full review instead of multiple single comments, but then I was already started

Reply:
> Sounds like the proposed code snippet above should be avoided.

That was the intension. Perhaps I should explicitly write "That's why this approach is *unsafe* *and should be avoided!*
And maybe in the introduction of the section:
"There is an alternative way to manually control allocation on a case to case basis. *This approach is not recommended but shown here for the sake of completeness*:"

Reply:
> Sorry for the multiple comments... Should have started a full review instead of multiple single comments, but then I was already started

No problem! Thanks for the review &ndash; it's always good to have a second pair of eyes looking at a large portion of text... :slightly_smiling_face: